### ISSUE TITLE
Distorted noise (playing at max volume) after skipping song and review/cue within song using librespot

### ISSUE THREAD

**Issue by twenta** [badaix/snapcast#1091]
Hi,

since a couple of month I get hammered by distorted noise playing at maximum volume when I skip a song and then play back or forth (hard to reproduce) playing songs in librespot (last release 0.4.2).

This is probably related to
https://github.com/badaix/snapcast/issues/735

Therefore I compiled the developer branch a few weeks ago, but this did not help.

Logs are not helpful, the snapclient message: 
**2023-01-05 13-37-23.614 [Info] (Stream) pMiniBuffer->full() && (abs(pMiniBuffer->mean()) > 50): -71810** is probably closed to the issue.

> 2023-01-05 13-36-37.455 [Info] (Alsa) PCM name: hw:CARD=CODEC,DEV=0, sample rate: 44100 Hz, channels: 2, buffer time: 80000 us, periods: 4, period time: 20000 us, period frames: 882
> 2023-01-05 13-36-46.043 [Info] (Stream) Exception: Not enough frames available, requested frames: 882, available: 754
> 2023-01-05 13-36-46.043 [Info] (Alsa) Failed to get chunk
> 2023-01-05 13-37-13.973 [Info] (Stream) abs(age > 500): 841000
> 2023-01-05 13-37-13.973 [Info] (Alsa) Failed to get chunk
> **2023-01-05 13-37-23.614 [Info] (Stream) pMiniBuffer->full() && (abs(pMiniBuffer->mean()) > 50): -71810**
> 2023-01-05 13-37-27.495 [Info] (Stream) Exception: Not enough frames available, requested frames: 882, available: 387
> 2023-01-05 13-37-27.495 [Info] (Alsa) Failed to get chunk
> 2023-01-05 13-37-32.501 [Notice] (Alsa) No chunk received for 5000ms. Closing ALSA.

I start librespot with the following params:

> source = librespot:///home/pi/.cargo/bin/librespot?name=Spotify&dryout_ms=500&username=###&password=###&devicename=Snapcast&bitrate=160&volume=70&normalize=true

**Comment by hamishfagg** [comment]
Do you mean that it plays white noise / static at max volume?
I have this issue too and I'm trying to work out that's happening exactly so I can report it.

I have a funny feeling that snapcast is getting out-of-sinc when reading from stdout from librespot.
I'll try to set up librespot and snapcast running separately over pipe or alsa loopback and see if it persists.

**Comment by twenta** [comment]
No, it really pushes the volume above any limit and that's why it sounds distorted. When I turn the volume control of the amplifier very low I can hear the music playing, but still distorted.

**Comment by hamishfagg** [comment]
Ok, different issue to mine then.

**Comment by j-lechner** [comment]
@hamishfagg @badaix on Snapcast v0.27.0 and librespot 0.4.2 I am facing the same (or at least similar) issue:

When changing tracks or cuing inside tracks on Spotify, sometimes it breaks down and just plays white noise. I can sporadically reproduce the issue by randomly changing tracks and cuing back/forth for a couple of minutes. 

Debug-level logs did not show anything suspicious. Tweaking various parameters including wd_timeout, sampleformat, chunk_ms did NOT solve the problem. 

What instead reliably works for me is integrating librespot via a pipe:
`mkfifo -m 777 /tmp/spotify`

`/usr/bin/librespot --name=Living --bitrate=320 --volume-ctrl=fixed --backend pipe --initial-volume=100 --device /tmp/spotify`

`source = pipe:///tmp/spotify?name=Spotify&mode=create&dryout_ms=2000&sampleformat=44100:16:2`

Hope this helps others.

**Comment by aidangoettsch** [comment]
Having a similar issue with a consistent reproduction. After pausing for enough time to get the message `No chunk received for 5000ms. Closing ALSA.` on the client, as well as at other random times, while playing Spotify through librespot, unpausing results in a very loud static noise. It does seem like there is some structure to the noise, as if it is playing the song playing at an out of bounds volume setting, so I think the theory here is correct. Pausing and playing again through Spotify makes audio go back to normal.

**Comment by badaix** [comment]
Sounds like librespot is writing something to stdout, which is not the raw PCM stream, e.g. a log message or something.

**Comment by fraintt** [comment]
Hi, I have the same issue that after few songs, the playback changes to a loud (white)noise. For me only restarting the SnapServer fixes the issue.

**Comment by oyvn** [comment]
I experience noise with this setup as well. I'm pretty sure it happens because my raspberry pi maxes its CPU (It's a RPI 1B). I don't know whether the noise comes from librespot or from snapcast though.

**Comment by badaix** [comment]
It would help a lot if you could also mention the version of librespot you're using. Last time I was on 0.2 and didn't have any problems. There were changes in the pipe packend, which is used by snapcast, which might cause the issue: https://github.com/librespot-org/librespot/commit/80020616fcfb27597917dee88a2bb9476bd090ee

If you have the problem with 0.4.2, please try to downgrade and if the issue disappears, open an issue at librespot.

**Comment by fraintt** [comment]
Sorry, you are right, I should have mentioned the versions. 
In my case according the apt history.log I upgraded librespot.v0.1.3 to librespot.v0.4.2

I have now downgraded librespot back to 0.4.1 and it seems to fix the issue, I will do further tests. 

At least in my case, it solved the problem.

**Comment by bitclick** [comment]
tldr:  workaround is using a named pipe as source instead of using ```source = librespot:///```

i think i am affected by this.  and i have new information:
- versions
  - librespot 0.4.2 618d7d9 (Built on 2023-04-14, Build ID: JMTNTZxV, Profile: release)
  - snapserver snapserver v0.27.0 (rev 54a3d862)
- i can reproduce this by seeking repeatedly in spotify and trying different songs.  takes me about a minute to trigger the bug
- the bug only occurs when running librespot through snapserver (`source = librespot:///...`), i cannot reproduce this when running librespot separately with a named pipe as fifo

i think this only occurs when having the audio piped directly (without using a named pipe). maybe snapserver confuses some logoutput with audio-data?  

is this an issue in librespot or could it be in snapserver?

**Comment by kingosticks** [comment]
There's this note at https://github.com/librespot-org/librespot/wiki/Audio-Backends#pipe

>  Skipping will shorten the current track and start a new Ogg 'stream' but the 'end of stream' flag for the skipped track will be missing. It is a slight non-conformity to Ogg format although all tested players accept it.

**Comment by manuelkue** [comment]
> @hamishfagg @badaix on Snapcast v0.27.0 and librespot 0.4.2 I am facing the same (or at least similar) issue:
> 
> When changing tracks or cuing inside tracks on Spotify, sometimes it breaks down and just plays white noise. I can sporadically reproduce the issue by randomly changing tracks and cuing back/forth for a couple of minutes.
> 
> Debug-level logs did not show anything suspicious. Tweaking various parameters including wd_timeout, sampleformat, chunk_ms did NOT solve the problem.
> 
> What instead reliably works for me is integrating librespot via a pipe: `mkfifo -m 777 /tmp/spotify`
> 
> `/usr/bin/librespot --name=Living --bitrate=320 --volume-ctrl=fixed --backend pipe --initial-volume=100 --device /tmp/spotify`
> 
> `source = pipe:///tmp/spotify?name=Spotify&mode=create&dryout_ms=2000&sampleformat=44100:16:2`
> 
> Hope this helps others.

Thanks that was the solution ❤
May I ask how you start the librespot-fifo reliabliy? Is there a way to automatically start it when the snapserver starts?
If I understood you correctly I have to 

- create the fifo: `mkfifo -m 777 /tmp/spotify`
- add the snapcast source in the /etc/snapserver.conf: `source = pipe:///tmp/spotify?name=Spotify&mode=create&dryout_ms=2000&sampleformat=44100:16:2`
- start to stream librespot to the fifo: `/usr/bin/librespot --name=Living --bitrate=320 --volume-ctrl=fixed --backend pipe --initial-volume=100 --device /tmp/spotify`

I'm a little confused because you noted the librespot part before creating the stream. Is there a smarter way than starting librespot manually and blocking the console, other than starting it from a new console and remembering to restart it when the snapserver restarts?

**Comment by j-lechner** [comment]
@manuelkue 

I've created a `/root/librespot.sh` script that also creates the fifo pipe (in case snapcast server starts only afterwards):

```
#!/bin/bash

mkfifo -m 777 /tmp/spotify
chown root:root /tmp/spotify
/usr/bin/librespot --name=Living --bitrate=320 --volume-ctrl=fixed --backend pipe --initial-volume=100 --device /tmp/spotify
```

which is automatically started and restarted via a `/etc/systemd/system/librespot.service`:

```
[Unit]
Description=librespot

[Service]
User=root
WorkingDirectory=/root
ExecStart=/root/librespot.sh
Restart=always

[Install]
WantedBy=multi-user.target
```


**Note: Bad idea to run this all as root, please change accordingly.**
